### PR TITLE
Update linkify to 4.1.1

### PR DESCRIPTION
**Pull request by Johennes**
This updates linkify to version 4.1.1. This version includes a backwards incompatible [refactoring of the token parsing](https://github.com/Hypercontext/linkifyjs/commit/e1237f7e441612e8da71c972aa80baf061a392fb) that required larger changes to our custom linkify plugins.

To be honest, I don't fully understand how our current plugins work. It looks like they're running through the same transition more than once, emitting tokens in _only some_ cases ... which doesn't match my understanding of state machines. This seems to somehow have been supported in previous linkify versions but now isn't anymore.

Fixes: vector-im/element-web#23806

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Update linkify to 4.1.1 ([\#11132](https://github.com/matrix-org/matrix-react-sdk/pull/11132)). Fixes vector-im/element-web#23806.

### CODE REVIEW

File: test/linkify-matrix-test.ts [hunk 1]
```diff
@@ -286,13 +311,21 @@ describe("linkify-matrix", () => {
             const found = linkify.find(test);
             expect(found).toEqual([
                 {
-                    href: char + "foo:bar.com-baz.com",
```
Comment (Johennes) [inline test/linkify-matrix-test.ts:289]: The test title says "does _not_ parse" but then the test asserted that the full string was accepted. This also doesn't seem to be a valid room identifier so I suspect this was a mistake?

**Comment by Johennes** [comment]
Hm, something is wrong with Cypress:

```
...
Still waiting to connect to Custom chromium, retrying in 1 second (attempt 60/62)
Still waiting to connect to Custom chromium, retrying in 1 second (attempt 61/62)
Still waiting to connect to Custom chromium, retrying in 1 second (attempt 62/62)
Cypress failed to make a connection to the Chrome DevTools Protocol after retrying for 50 seconds.
```

🤔

File: src/linkify-matrix.ts [hunk 2]
```diff
@@ -231,23 +217,21 @@ export const options: Opts = {
...
         scanner,
         parser,
-        utils,
         token,
         name: Type.RoomAlias,
     });
 });
 
-registerPlugin(Type.UserId, ({ scanner, parser, utils }: any) => {
+registerPlugin(Type.UserId, ({ scanner, parser }: any) => {
```
Comment (t3chguy) [inline src/linkify-matrix.ts:230]: Can we ditch the `any`

File: test/linkify-matrix-test.ts (hunk 1 above)
Comment (t3chguy) [reply test/linkify-matrix-test.ts:289]: Why would `#foo:bar.com-baz.com` not be a valid room identifier? `com-baz.com` looks like a valid domain, `bar` being a valid subdomain. Dashed in domains are valid, looks like this test being updated asserts they are not. https://www.fasthosts.co.uk/blog/hyphens-in-domain-names/

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: test/linkify-matrix-test.ts (hunk 1 above)
Comment (Johennes) [reply test/linkify-matrix-test.ts:289]: Hm, yeah I think you're right. Hyphens are also allowed according to https://spec.matrix.org/v1.7/appendices/#server-name. But then the name of the test is sort of misleading?

It's also interesting that linkify's `domain` group doesn't appear to include hyphens.

Comment (t3chguy) [reply test/linkify-matrix-test.ts:289]: > But then the name of the test is sort of misleading?

Agreed

> It's also interesting that linkify's domain group doesn't appear to include hyphens.

Quite

Comment (Johennes) [reply test/linkify-matrix-test.ts]: Have re-introduced the additional domainpart tokens to fix this

**Review by t3chguy — APPROVED** [review]
:shipit:

**Pull request merged**